src/exact.jl

```diff
@@ -0,0 +1,319 @@
+end
+
+"""
+    ot_cost(
+        c, μ::ContinuousUnivariateDistribution, ν::UnivariateDistribution; plan=nothing
+    )
+
+Compute the optimal transport cost for the Monge-Kantorovich problem with univariate
+distributions `μ` and `ν` as source and target marginals and cost function `c` of
+the form ``c(x, y) = h(|x - y|)`` where ``h`` is a convex function.
+
+In this setting, the optimal transport cost can be computed as
+```math
+\\int_0^1 c(F_\\mu^{-1}(x), F_\\nu^{-1}(x)) \\mathrm{d}x
+```
+where ``F_\\mu^{-1}`` and ``F_\\nu^{-1}`` are the quantile functions of `μ` and `ν`,
+respectively.
+
+A pre-computed optimal transport `plan` may be provided.
+
+See also: [`ot_plan`](@ref), [`emd2`](@ref)
+"""
+function ot_cost(
+    c, μ::ContinuousUnivariateDistribution, ν::UnivariateDistribution; plan=nothing
+)
+    cost, _ = if plan === nothing
+        quadgk(0, 1) do q
+            return c(quantile(μ, q), quantile(ν, q))
+        end
+    else
+        quadgk(0, 1) do q
+            x = quantile(μ, q)
+            return c(x, plan(x))
+        end
+    end
+    return cost
+end
+
+################
+# Discrete 1D OT
+################
+
+# internal iterator for discrete one-dimensional OT problems
+# it returns tuples that consist of the indices of the source and target histograms
+# and the optimal flow between the corresponding points
+struct Discrete1DOTIterator{T,M,N}
+    mu::M
+    nu::N
+end
+
+# histograms `μ` and `ν` are expected to be iterators of the histograms where the
+# corresponding support is sorted
+function Discrete1DOTIterator(μ, ν)
+    T = Base.promote_eltype(μ, ν)
+    return Discrete1DOTIterator{T,typeof(μ),typeof(ν)}(μ, ν)
+end
+
+Base.IteratorEltype(::Type{<:Discrete1DOTIterator}) = Base.HasEltype()
+Base.IteratorSize(::Type{<:Discrete1DOTIterator}) = Base.SizeUnknown()
+Base.eltype(::Type{<:Discrete1DOTIterator{T}}) where {T} = Tuple{Int,Int,T}
+
+# we iterate through the source and target histograms
+function Base.iterate(
+    d::Discrete1DOTIterator{T}, (i, j, μnext, νnext)=(1, 1, iterate(d.mu), iterate(d.nu))
+) where {T}
+    # if we are done with iterating through the source and/or target histogram,
+    # iteration is stopped
+    if μnext === nothing || νnext === nothing
+        return nothing
+    end
+
+    # unpack next values and states of the source and target histograms
+    μiter, μstate = μnext
+    νiter, νstate = νnext
+
+    # compute next value of the iterator: indices of source and target histograms
+    # and optimal flow between the corresponding points
+    min_iter, max_iter = minmax(μiter, νiter)
+    iter = (i, j, min_iter)
+
+    # compute next state of the iterator
+    diff = max_iter - min_iter
+    state = if μiter < max_iter
+        # move forward in the source histogram
+        (i + 1, j, iterate(d.mu, μstate), (diff, νstate))
+    else
+        # move forward in the target histogram
+        (i, j + 1, (diff, μstate), iterate(d.nu, νstate))
+    end
+
+    return iter, state
+end
```

Comment:
Does creating an iterator improves the speed of the code?

Reply:
The iterator per se does not improve performance, it is the same logic as in the explicit implementation. However, it removes the type instability and allows one to compute the cost without ever allocating an array for the plan in an easy way.